feat(v2): docusaurus-plugin-client-redirects

slorber opened the pull request:
First draft of plugin to handle client side redirects

lex111 commented on packages/docusaurus-plugin-client-redirects/src/createRedirectPageContent.ts:
```diff
@@ -0,0 +1,35 @@
+  toUrl,
+}: CreateRedirectPageOptions) {
+  return `
+<!DOCTYPE html>
```
How about using Eta to render an HTML template from separated file? You can find an example below: https://github.com/facebook/docusaurus/blob/368eb4279ff6a1f26d134a2349e2ebd8d9c35644/packages/docusaurus/src/client/serverEntry.js#L64

slorber replied:
done

lex111 commented on packages/docusaurus-plugin-client-redirects/src/index.ts:
```diff
@@ -0,0 +1,144 @@
+async function writeRedirectFiles(redirects: RedirectMetadata[]) {
+  async function writeRedirectFile(redirect: RedirectMetadata) {
+    try {
+      await fs.writeFile(
```
Please use sync file write (see https://github.com/facebook/docusaurus/pull/2530 for more details and examples).

slorber replied:
This shouldn't be a problem in this case, because I'm using fs-extra and awaiting the file writes

lex111 replied:
Alright, if so, otherwise in certain cases the website building may be finished before the redirect files are generated.

lex111 commented on packages/docusaurus-plugin-client-redirects/src/redirectCreators.ts:
```diff
@@ -0,0 +1,41 @@
+    const extensionMatch = dottedExtensions.find((ext) =>
+      fromRoutePath.endsWith(`.${ext}`),
+    );
+    if (extensionMatch) {
```
I would change the `extensionMatch` const name to something more understandable, maybe `foundExtension`?

slorber replied:
done

lex111 commented on packages/docusaurus-plugin-client-redirects/src/utils.ts:
```diff
@@ -0,0 +1,23 @@
+}
+
+export function removeTrailingSlash(str: string) {
+  return str.endsWith('/') ? str.substr(0, str.length - 2) : str;
```
Could we do this?
`str.substr(0, str.length - 2)` -> `str.slice(0, -1)`

It should be faster.

slorber replied:
ah yes, always forget we can write this :)

slorber commented:
Thanks for the feedback @lex111 

It's not ready and not tested, just wanted to let you see a preview of the API I'm crafting (actually didn't even run the code yet so it might have bugs)

Will handle your feedbacks tomorrow.

slorber commented on packages/docusaurus-plugin-client-redirects/src/createRedirectPageContent.ts:
```diff
@@ -0,0 +1,35 @@
+      document.write('If you are not redirected automatically, follow this <a href="' + redirectLink + '">link</a>.')
+      document.write('</body>')
+      setTimeout(() => {
+        window.location.assign(redirectLink)
```
@lex111 this part is inspired by the plugin you wrote here: https://github.com/single-spa/single-spa.js.org/blob/master/website/src/plugins/docusaurus-plugin-redirects/src/index.js

Was wondering why do you need a setTimeout?

I've seen a gatsby plugin that does not have this timeout

https://github.com/dhakerShiv/gatsby-plugin-client-side-redirect/blob/master/clientSideRedirect.js

lex111 replied:
Oh yes, this is an  old-fashion version of browser redirect. We had better remove `setTimeout` and `document.write` calls.

slorber replied:
done

slorber commented:
Hi @lex111 

I've added the tests and made the code more robust, it's almost finished (still has some bugs to fixes, and the write side involving fs is not tested yet).

When this feature is complete, using the following (main usecase)

```
    [
      '@docusaurus/plugin-client-redirects',
      {
        fromExtensions: ['html'],
      },
    ],
```

Should redirect from "/path.html" to "/path" and write a "/path.html/index.html" file (or maybe just "/path.html" ?)
 

But it should also support the "Watchman" case (https://github.com/facebook/docusaurus/issues/2697) that wants .html urls being the canonical ones:

```
    [
      '@docusaurus/plugin-client-redirects',
      {
        toExtensions: ['html'],
      },
    ],
```

Should redirect from "/path" to "/path.html" and write a "/path/index.html" file


For power users, it's also possible to provide a function:

```
    [
      '@docusaurus/plugin-client-redirects',
      {
        createRedirects: (existingPath) => {
           return ["existingPath" + "/" + "someFancyPathSuffix"];
        },
      },
    ],
```

slorber commented:
So I think it's working fine now, and has good test coverage.

I propose to dogfood on the docusaurus v2 to redirect /path.html to /path, so that we know it keeps working over time.

It's already in the PR: https://deploy-preview-2793--docusaurus-2.netlify.app/docs/next/installation.html

I'll try the plugin on the Watchman site now and see if I have any issue.

And I'll write the doc.

slorber commented on packages/docusaurus-plugin-client-redirects/src/__tests__/writeRedirectFiles.test.ts:
```diff
@@ -0,0 +1,116 @@
+    );
+
+    expect(redirectFiles.map((f) => f.fileAbsolutePath)).toEqual([
+      path.join(pluginContext.outDir, '/abc.html/index.html'),
```
@lex111 @yangshun , this is some part I'm not totally sure of.

If we need to redirect from path `/abc.html`, should we write `/abc.html` to the disk or `/abc.html/index.html` ? Is there a method that already converts a route path to a fs file path somewhere?

slorber commented on packages/docusaurus-plugin-client-redirects/src/__tests__/writeRedirectFiles.test.ts:
```diff
@@ -0,0 +1,116 @@
+
+    await expect(writeRedirectFiles(filesMetadata)).rejects.toThrowError(
+      `Redirect file creation error for path=${filesMetadata[0].fileAbsolutePath}: Error: The redirect plugin is not supposed to override existing files`,
+    );
```
@lex111 @yangshun  this is something I guess is important: no matter the user config, we should never allow a redirect file to override an existing file that was emitted for an existing path

slorber commented on packages/docusaurus-plugin-client-redirects/src/collectRedirects.ts:
```diff
@@ -0,0 +1,89 @@
+    (redirect) => !pluginContext.routesPaths.includes(redirect.fromRoutePath),
+  );
+
+  return redirects;
```
@lex111 @yangshun is there a decent way to warn the user in case conflicting redirects are found?

For example when we detect one of his redirect rules would override an existing path, or if he ends up having twice the same redirect or something?

Nothing prevents the user to do weird things like `createRedirects(originalPath) => [originalPath + ".html",originalPath + ".html"]`

For now I just silently "filter" those to avoid overriting existing route file, or writing twice the same redirect, but maybe it should warn the user instead?

slorber commented on packages/docusaurus-plugin-client-redirects/src/normalizePluginOptions.ts:
```diff
@@ -0,0 +1,45 @@
+  return value instanceof Function;
+}
+
+const UserOptionsSchema = Yup.object().shape<UserPluginOptions>({
```
I've included Yup because I think it's a good solution to validate unsafe user inputs like plugin options. It gives a way to show a better error to the user instead of failing late with a cryptic runtime error

slorber commented on packages/docusaurus-plugin-client-redirects/src/templates/redirectPage.template.html.ts:
```diff
@@ -0,0 +1,20 @@
+  <head>
+    <meta charset="UTF-8">
+    <meta http-equiv="refresh" content="0; url=<%= it.toUrl %>">
+    <link rel="canonical" href="<%= it.toUrl %>" />
```
I wonder if it's useful to include a robots noIndex here? Not sure this is very useful but if you know...

slorber commented on packages/docusaurus-plugin-client-redirects/src/utils.ts:
```diff
@@ -0,0 +1,30 @@
+  const fileName = path.basename(routePath);
+  const filePath = path.dirname(routePath);
+  return path.join(filePath, `${fileName}/index.html`);
+}
```
This function may not be perfect, or already exist somewhere, any idea? @lex111 @yangshun ?

slorber commented on packages/docusaurus-plugin-client-redirects/src/__tests__/redirectCreators.test.ts:
```diff
@@ -0,0 +1,87 @@
+    expect(redirectCreator('')).toEqual([]);
+    expect(redirectCreator('/')).toEqual([]);
+    expect(redirectCreator('/abc')).toEqual(['/abc.html', '/abc.htm']);
+    expect(redirectCreator('/def.html')).toEqual([]);
```
@yangshun @lex111 I'm not totally sure of this behavior:

If user wants to redirect from html + htm extensions, and the route already ends with one of the extensions he wants to redirect from (like `/def.html` here), I guess It's not worth creating redirections from path `/def.html.html` and `/def.html.htm`, that would be weird.

It's quite unlikely to happen and probably doesn't matter but if you have an opinion I'm curious

slorber commented on packages/docusaurus-plugin-client-redirects/src/writeRedirectFiles.ts:
```diff
@@ -0,0 +1,86 @@
+      `${removeTrailingSlash(pluginContext.baseUrl)}${path.join(
+        redirect.toRoutePath,
+      )}`,
+    );
```
@yangshun @lex111 I'm not totally sure either if always having a trailing slash in redirect urls is good and supported by all host providers, I guess it's related to the FS redirect file location question as well.

lex111 commented:
I wonder, in the current Docusaurus architecture, is it possible to define redirects in the frontmatter of Markdown files, like in this Gatsby plugin? https://www.gatsbyjs.org/packages/gatsby-redirect-from/

slorber commented:
Hi @lex111 

I definitively agree that it's convenient to declare redirects directly from the markdown files.

Gatsby's architecture is a bit different than D2. It has a data layer that permits to read the frontmatter of all markdown files from everywhere, and a core `createRedirect` api that they leverage mostly to hook into host specific solutions, and create server-side redirects (for example, if you use gatsby-plugin-netlify, it will write a _redirects file to the /build folder, which is a proprietary config file for Netlify).

In D2 we don't really have a shared data layer to read about all markdown files from a unified interface. Instead, each plugin is independant, reads frontmatter of the markdown on his own, and do the calls to `addRoute`. The frontmatter for a blog post and a doc are different types. 


Even with a core `createRedirect` api, we would have to implement the reading of `redirect_from: []` in each markdown plugin to call that `createRedirect` api, while gatsby is able to query its data layer directly in a more generic way.


Also, implementing redirects at the markdown level means it's not possible to create redirects from JSX pages.

I think it's better to start with this generic plugin. The configuration feels less "co-located" and more centralized than the Gatsby one, but at least it should support all usecases (ie markdown + pages), without the need to add new core lifecycle actions like `createRedirect`. 

Also we should discourage users to use client-side redirects, in favor of server-side redirects. As we don't want to support host-specific redirect configuration, introducing a core `createRedirect` api would encourage people to create client-side redirects. 

I see this plugin as an escape hatch when the user can't setup server-side redirects at all for various reasons (like using Github pages and not willing to migrate, which is the Watchman case).

slorber commented on website/docs/migrating-from-v1-to-v2.md:
````diff
@@ -420,6 +420,43 @@ The following fields are all deprecated, you may remove from your configuration
+```
+
+If you want to keep the `.html` extension as the canonical url of a page, docs can declare a `slug: installation.html` frontmatter.
+
````
@lex111 FYI I added some migration doc about `cleanUrl`

lex111 commented on website/docs/using-plugins.md:
```diff
@@ -458,3 +458,78 @@ import thumbnail from './path/to/img.png';
+
+Docusaurus Plugin to generate **client-side redirects**.
+
+This plugin will write additional HTML pages to your static site, that redirects the user to your existing Docusaurus pages with Javascript.
```
```suggestion
This plugin will write additional HTML pages to your static site, that redirects the user to your existing Docusaurus pages with JavaScript.
```

lex111 commented on website/docs/using-plugins.md:
```diff
@@ -458,3 +458,78 @@ import thumbnail from './path/to/img.png';
+
+Before using this plugin, you should look if your hosting provider doesn't offer this feature.
+
+As far as we know, only Github Pages does not offer server-side redirects.
```
I think it’s better to remove this sentence, as it is quite specific. In the future, GitHub Pages may implement redirect support, and we will forget to update the docs (just an example).

slorber replied:
@lex111 only the last one right? agree

lex111 replied:
Yes

slorber commented:
> An alternative is to be something like this:
> 
> ```
>   additionalRedirects: [
>     {
>       to: '/docs/newDocPath',
>       from: ['/docs/oldDocPathFrom2019', '/docs/legacyDocPathFrom2016'],
>     },
>     {
>       to: '/docs/newDocPath',
>       from: ['/docs/oldDocPathFrom2019', '/docs/legacyDocPathFrom2016'],
>     },
>     {
>       to: '/docs/newDocPath',
>       from: ['/docs/oldDocPathFrom2019', '/docs/legacyDocPathFrom2016'],
>     },
>   ],
> ```

Yes, I'll add declaring redirects as an array too!


> By the way, is it possible to use `fromExtensions`/`toExtensions` and `createRedirects` **together**? In one case, I added redirects in migration purposes, but later I needed to make a new redirect for a specific page (for example, I refactored docs a little bit).


Yes, we can use all options at the same time.

slorber commented:
@lex111 currently I have:

```js
module.exports = {
  plugins: [
    [
      "@docusaurus/plugin-client-redirects",
      {
        fromExtensions: ["html"],
        toExtensions: ["exe"],
        redirects: [
          {
            to: "/docs/newDocPath",
            from: ["/docs/oldDocPathFrom2019", "/docs/legacyDocPathFrom2016"]
          }
        ],
        createRedirects: function(existingPath) {
          if (existingPath === "/docs/newDocPath") {
            return ["/docs/oldDocPathFrom2019", "/docs/legacyDocPathFrom2016"];
          }
        }
      }
    ]
  ]
};
```

I'm thinking on making the `createRedirects` function more consistent, what do you think of:

```js
module.exports = {
  plugins: [
    [
      "@docusaurus/plugin-client-redirects",
      {
        fromExtensions: ["html"],
        toExtensions: ["exe"],
        redirects: [
          {
            to: "/docs/newDocPath",
            from: ["/docs/oldDocPathFrom2019", "/docs/legacyDocPathFrom2016"]
          }
        ],
        createRedirects: function({paths, createRedirect}) {
              actions.createRedirect({
                from: "/docs/oldDocPathFrom2019",
                to: "/docs/newDocPath"
              });
              actions.createRedirect({
                from: "/docs/legacyDocPathFrom2016",
                to: "/docs/newDocPath"
              });
          });
        }
      }
    ]
  ]
};
```

It's a bit more verbose, but more consistent with the static redirect array you proposed and could be easier to understand as well.

Also it's what Gatsby does and nobody complained afaik: https://www.gatsbyjs.org/docs/actions/#createRedirect

lex111 commented:
> I'm thinking on making the createRedirects function more consistent, what do you think of:

I don’t know what are the advantages of this approach?

By the way, after the merging of #2861, it would be nice to add redirects to v2 website. I wonder if it is possible to use regular expressions for this purpose (to cover versioned docs)?


https://v2.docusaurus.io/docs/introduction -> https://v2.docusaurus.io/docs/
https://v2.docusaurus.io/docs/2.0.0-alpha.55/introduction -> https://v2.docusaurus.io/docs/2.0.0-alpha.55/
https://v2.docusaurus.io/docs/2.0.0-alpha.54/introduction -> https://v2.docusaurus.io/docs/2.0.0-alpha.54/
https://v2.docusaurus.io/docs/2.0.0-alpha.50introduction -> https://v2.docusaurus.io/docs/2.0.0-alpha.50/
https://v2.docusaurus.io/docs/2.0.0-alpha.48/introduction -> https://v2.docusaurus.io/docs/2.0.0-alpha.48/
https://v2.docusaurus.io/docs/next/introduction -> https://v2.docusaurus.io/docs/next/


I did something similar when [I needed to correctly highlight active items in navbar](https://github.com/facebook/docusaurus/blob/master/website/docusaurus.config.js#L103), so I'm  curious if redirects could be added this way (using regex).

slorber commented:
Hi @lex111 

Here are some updates:

- a new `redirects` plugin option, as you requeted
- improved documentation an example with the 4 plugin options used together
- stricter validation: it's barely impossible for the user to use badly formatted options
- improved error messages
- non-blocking warnings in case some redirects get filtered for security (ie that would lead to FS files overwriting each thers)
- did not change the `createRedirects` (cf comment above)


----

Here's the config deployed in the PR, so that we can test better on D2 website:

```js
[
      '@docusaurus/plugin-client-redirects',
      {
        fromExtensions: ['html'],
        redirects: [
          {
            to: '/',
            from: [
              '/plugin-client-redirects-tests/toHome1',
              '/plugin-client-redirects-tests/toHome2',
            ],
          },
          {
            to: '/docs',
            from: '/plugin-client-redirects-tests/toDocs1',
          },
          {
            to: '/docs',
            from: '/plugin-client-redirects-tests/toDocs2',
          },
          {
            to: '/docs',
            from: '/plugin-client-redirects-tests/toHomeDuplicatePath',
          },
        ],
        createRedirects: function (existingPath) {
          if (existingPath === '/') {
            return [
              '/',
              '/docs',
              '/plugin-client-redirects-tests/toHome3',
              '/plugin-client-redirects-tests/toHome4',
              '/plugin-client-redirects-tests/toHomeDuplicatePath',
            ];
          }
        },
      },
    ],
```

The following redirects work:

Html extension redirects:

- https://deploy-preview-2793--docusaurus-2.netlify.app/docs/next.html
- https://deploy-preview-2793--docusaurus-2.netlify.app/docs/next/installation.html

Other redirects:

- https://deploy-preview-2793--docusaurus-2.netlify.app/plugin-client-redirects-tests/toHome1
- https://deploy-preview-2793--docusaurus-2.netlify.app/plugin-client-redirects-tests/toHome2
- https://deploy-preview-2793--docusaurus-2.netlify.app/plugin-client-redirects-tests/toHome3
- https://deploy-preview-2793--docusaurus-2.netlify.app/plugin-client-redirects-tests/toHome4
- https://deploy-preview-2793--docusaurus-2.netlify.app/plugin-client-redirects-tests/toHomeDuplicatePath
- https://deploy-preview-2793--docusaurus-2.netlify.app/plugin-client-redirects-tests/toDocs1
- https://deploy-preview-2793--docusaurus-2.netlify.app/plugin-client-redirects-tests/toDocs2

As the config creates twice the `toHomeDuplicatePath`, and tries to redirect from `/` and `/docs` (existing paths that should not be overwritten by a redirect file), the build outputs the following warnings:

![image](https://user-images.githubusercontent.com/749374/83671109-70a91380-a5d4-11ea-93bf-141c597a1874.png)


Do you see other things to handle?

slorber commented:
About the version thing, I thought about it, and agree that's something we could support. 
What about adding this in a new PR? As this one starts to get fat :)
It's already possible to add version redirects with the custom `createRedirects` fn in userland.
But I agree this is a pattern that we should make it simpler to configure.

lex111 commented:
Honestly, this plugin really came out huge and strong I thought there would be much less code in order to implement it. We actually did not pay so much attention to error handling in our existing plugins. However, I can’t say that this is a bad approach, just before that we did not attach so much importance to this.

I actually trust in a high test coverage of this plugin, since mentally this plugin is not easy to understand (I mean me). But this is really cool, although maybe we could simplify some things, for example, is it really necessary to use Yup? Although it may be necessary to go this way in our other plugins. I mean, maybe we should try to reconsider the development process of our plugins so that they are like this redirect plugin -- a lot of tests, a thorough validation of errors, a lot of high-level abstractions... Apparently I got carried away :smile: 

Okay, tomorrow I'll see it again in action and probably we can already merge it into master branch.

P.S. But please add redirects for `introduction` path. This will also be an example for release notes.

slorber commented:
Hi @lex111 

> Honestly, this plugin really came out huge and strong I thought there would be much less code in order to implement it.

That's true, initially thought it would be smaller 😅

The size comes from giving more control to the user, and giving shortcuts for common patterns, with the fromExtensions/toExtensions.


> We actually did not pay so much attention to error handling in our existing plugins. However, I can’t say that this is a bad approach, just before that we did not attach so much importance to this.

> for example, is it really necessary to use Yup?

> Although it may be necessary to go this way in our other plugins. I mean, maybe we should try to reconsider the development process of our plugins so that they are like this redirect plugin -- a lot of tests, a thorough validation of errors, a lot of high-level abstractions...

That's true. I did more user input validation in this plugin than other plugins. Personally I'd rather validate all user inputs on all plugins, fail-fast, and give actionable error messages to the user whenever possible, as it would lead to less github issues on the long term. 

If you feel it's too much, not worth it, and not the direction we want to take, I'm ok with that and could remove all these validations.

> since mentally this plugin is not easy to understand (I mean me)

> a lot of high-level abstractions

I agree that it's not so easy to reason about this plugin, there are more hidden subtilities than I thought. Myself I often get confused, particularly inverting the from/to of the plugin. Maybe it's not the correct terminology to use, and we could find simpler wordings (like sourcePath/targetPath or something?)

About the abstractions, I don't think there's a lot of them, the only thing I really abstracted is the "redirectCreator" concept, which is `function(path): redirects`, and permit to fit the user provided `createRedirects` function and the "shortcuts" `fromExtensions` / `toExtensions` under the same interface.

I can remove this abstraction if you feel it's too hard to reason about. That means I'll need to write less generic code here:

```js
function doCollectRedirects(pluginContext: PluginContext): RedirectMetadata[] {
  const redirectsCreators: RedirectsCreator[] = buildRedirectCreators(
    pluginContext.options,
  );

  const optionsRedirects = collectPluginOptionRedirects(pluginContext);

  const redirectCreatorsRedirects = flatten(
    redirectsCreators.map((redirectCreator) => {
      return createRoutesPathsRedirects(redirectCreator, pluginContext);
    }),
  );

  return [...optionsRedirects, ...redirectCreatorsRedirects];
}
```

That would become something like:

```js
function doCollectRedirects(pluginContext: PluginContext): RedirectMetadata[] {
  return [
    ...collectToExtensionsRedirects(pluginContext),
    ...collectFromExtensionsRedirects(pluginContext),
    ...collectPluginOptionRedirects(pluginContext),
    ...collectCreateRedirects(pluginContext),
  ];
}
```

I think you are right, that may be simpler to understand, and this abstraction may not be worth it.

Do you agree I should refactor?

slorber commented on packages/docusaurus-plugin-client-redirects/src/collectRedirects.ts:
```diff
@@ -0,0 +1,168 @@
+      pluginContext.options.createRedirects,
+    ),
+  ];
+}
```
@lex111 after removal of createRedirects abstraction

slorber commented:
@lex111 

I've removed the "redirectsCreator" abstraction, that simplified a bit the code.


For the doc home redirects, blocked by https://github.com/facebook/docusaurus/issues/2879
Maybe we can merge and fix this redirection later if needed?

yangshun commented on website/docusaurus.config.js:
```diff
@@ -21,6 +21,44 @@ module.exports = {
+      '@docusaurus/plugin-client-redirects',
+      {
+        fromExtensions: ['html'],
+        redirects: [
```
Should we remove all these before merging?
